src/platform_linux/shmem.c

```diff
@@ -0,0 +1,504 @@
+bool Trace_shmem        = FALSE;
+
+/*
+ * -----------------------------------------------------------------------------
+ * Core structure describing shared memory segment created. This lives right
+ * at the start of the allocated shared segment.
+ * -----------------------------------------------------------------------------
+ */
+typedef struct shmem_info {
+   void  *shm_start;       // Points to start address of shared segment.
+   void  *shm_next;        // Points to next 'free' address to allocate from.
+   size_t shm_total_bytes; // Total size of shared segment allocated initially.
+   size_t shm_free_bytes;  // Free bytes of memory left (that can be allocated)
+   size_t shm_used_bytes;  // Used bytes of memory left (that were allocated)
+   uint64 shm_magic;       // Magic identifier for shared memory segment
+   int    shm_id;          // Shared memory ID returned by shmget()
+
+} PLATFORM_CACHELINE_ALIGNED shmem_info;
+
+/* Permissions for accessing shared memory and IPC objects */
+#define PLATFORM_IPC_OBJS_PERMS (S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP)
+
+/*
+ * PLATFORM_HEAP_ID_TO_HANDLE() --
+ * PLATFORM_HEAP_ID_TO_SHMADDR() --
+ *
+ * The shared memory create function returns the address of shmem_info->shm_id
+ * as the platform_heap_id heap-ID to the caller. Rest of Splinter will use this
+ * heap-ID as a 'handle' to manage / allocate shared memory. This macro converts
+ * the heap-ID handle to the shared memory's start address, from which the
+ * location of the next-free-byte can be tracked.
+ *
+ * RESOLVE - This ptr-math going back from address of heap-ID is prone to
+ *errors. In some code paths, e.g. tests/unit/btree_test.c, we pass-down the
+ *stack address of an on-stack scratch-buffer masquerading as the heap-ID
+ *handle.
+ */
+static inline platform_heap_handle
+platform_heap_id_to_handle(platform_heap_id hid)
+{
+   return (platform_heap_handle)((void *)hid - offsetof(shmem_info, shm_id));
+}
+
+static inline void *
+platform_heap_id_to_shmaddr(platform_heap_id hid)
+{
+   debug_assert(hid != NULL);
+   void *shmaddr = (void *)platform_heap_id_to_handle(hid);
+   return shmaddr;
+}
+
+/*
+ * -----------------------------------------------------------------------------
+ * platform_shmcreate() -- Create a new / attach to an existing shared segment.
+ *
+ * For a given set of heap ID/handle, we expect that this create method will
+ * only be called once. [ Otherwise, it means some code is erroneously creating
+ * the shared segment twice, clobbering previously established handles. ]
+ * -----------------------------------------------------------------------------
+ */
+platform_status
+platform_shmcreate(size_t                size,
+                   platform_heap_handle *heap_handle,
+                   platform_heap_id     *heap_id)
+{
+   platform_assert(*heap_handle == NULL);
+   platform_assert(*heap_id == NULL);
+
+   int shmid = shmget(0, size, (IPC_CREAT | PLATFORM_IPC_OBJS_PERMS));
```

Comment:
`shmget` is the older System V shared memory API.  For new codebases, the man pages recommend using POSIX shared memory.  See the "Notes" section here: https://man7.org/linux/man-pages/man7/shm_overview.7.html

Reply:
In particular, if I'm reading the man pages correctly, a POSIX shared memory object may be dynamically resized using a combination of `ftruncate` and `mremap`.  That may be useful here.

Reply:
@rosenhouse -- Thanks for giving this a look. As I said earlier, this is a very early draft change set. I got something working and put it out there for discussion. 

I'll look into your Posix-suggestions later on ... once I get this cobbled together with Postgres.